common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,893 @@
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(appId, shuffleId, reduceId);
+    }
+
+    @Override
+    public String toString() {
+      return Objects.toStringHelper(this)
+        .add("appId", appId)
+        .add("shuffleId", shuffleId)
+        .add("reduceId", reduceId)
+        .toString();
+    }
+
+    String generateFileName() {
+      return String.format("mergedShuffle_%s_%d_%d", appId, shuffleId, reduceId);
+    }
+
+    boolean compareAppShuffleId(String appId, int shuffleId) {
+      return Objects.equal(this.appId, appId) && this.shuffleId == shuffleId;
+    }
+  }
+
+  /** Metadata tracked for an actively merged shuffle partition */
+  public static class AppShufflePartitionInfo {
+
+    private final AppShufflePartitionId partitionId;
+    // The merged shuffle data file
+    final File dataFile;
+    public final FileChannel channel;
+    // Location offset of the last successfully merged block for this shuffle partition
+    private long position;
+    // Indicating whether failure was encountered when merging the previous block
+    private boolean encounteredFailure;
+    // Track the map Id whose block is being merged for this shuffle partition
+    private int currentMapId;
+    // Bitmap tracking which mapper's blocks have been merged for this shuffle partition
+    private RoaringBitmap mapTracker;
+    // The index file for a particular merged shuffle contains the chunk offsets.
+    private final FileChannel indexChannel;
+    // The meta file for a particular merged shuffle contains all the map ids that belong to every
+    // chunk. The entry per chunk is a serialized bitmap.
+    private final FileChannel metaChannel;
+    private final DataOutputStream indexWriteStream;
+    // The offset for the last chunk tracked in the index file for this shuffle partition
+    private long lastChunkOffset;
+    private int lastMergedMapId = -1;
+
+    // Bitmap tracking which mapper's blocks are in the current shuffle chunk
+    private RoaringBitmap chunkTracker;
+    ByteBuf trackerBuf = null;
+
+    AppShufflePartitionInfo(
+        AppShufflePartitionId partitionId,
+        File dataFile,
+        File indexFile,
+        File metaFile) throws IOException {
+      this.partitionId = Preconditions.checkNotNull(partitionId, "partition id");
+      dataFile.createNewFile();
+      this.dataFile = dataFile;
+      this.channel = new FileOutputStream(dataFile, true).getChannel();
+      indexFile.createNewFile();
+      FileOutputStream fos = new FileOutputStream(indexFile, true);
+      indexChannel = fos.getChannel();
+      this.indexWriteStream = new DataOutputStream(new BufferedOutputStream(fos));
+      metaFile.createNewFile();
+      metaChannel = new FileOutputStream(metaFile, true).getChannel();
+      this.currentMapId = -1;
+      // Writing 0 offset so that we can reuse ShuffleIndexInformation.getIndex()
+      updateChunkInfo(0L, -1);
```

Comment:
Can't we simply do `indexWriteStream.writeLong(0L)` here? I prefer to get rid of the special case of `mapId=-1` here.

Reply:
Will have to repeat some log statements. If there is an exception just while writing to index file, we do want the logs to reflect it. So, some trivial code is repeated. I do prefer writing to index file from a single method instead of writing it from different locations. I think that helps with debugging.
Anyways, I don't have a strong opinion about this. If you insist on changing this, I will make that change.